File: folkmq-lang-js/folkmq/src/client/MqMessageReceived.ts
import {MqClientInternal} from "./MqClient";
import {Message} from "@noear/socket.d/transport/core/Message";
import {Session} from "@noear/socket.d/transport/core/Session";
import {MqConstants} from "../common/MqConstants";
import {Entity} from "@noear/socket.d/transport/core/Entity";
import {MqUtils} from "../common/MqUtils";
import {MqMessageBase} from "./MqMessage";

export interface MqMessageReceived extends MqMessageBase{
    /**
     * 主题
     */
    getTopic(): string;

    /**
     * 内容
     *
     * @deprecated 1.4
     * */
    getContent(): string;

    /**
     * 内容
     */
    getBodyAsString(): string;

    /**
     * 消费者组
     */
    getConsumerGroup(): string;

    /**
     * 已派发次数
     */
    getTimes(): number;

    /**
     * 回执
     */
    acknowledge(isOk: boolean);

    /**
     * 响应
     * */
    response(entity: Entity | null);
}

export class MqMessageReceivedImpl implements MqMessageReceived {
    private _clientInternal: MqClientInternal;
    private _source: Message;
    private _session: Session;

    private readonly _sender: string;
    private readonly _tid: string;
    private readonly _tag: string;
    private readonly _topic: string;
    private readonly _consumerGroup: string;
    private readonly _expiration: Date | null;
    private readonly _sequence: boolean;
    private readonly _transaction: boolean;
    private readonly _qos: number;
    private readonly _times: number;

    constructor(clientInternal: MqClientInternal, session: Session, source: Message) {
        this._clientInternal = clientInternal;
        this._session = session;
        this._source = source;

        let mr = MqUtils.getOf(source);

        this._sender = mr.getSender(source);

        this._tid = mr.getTid(source);
        this._tag = mr.getTag(source);
        this._topic = mr.getTopic(source);
        this._consumerGroup = mr.getConsumerGroup(source);

        this._qos = mr.getQos(source);
        this._times = mr.getTimes(source);
        this._sequence = mr.isSequence(source);
        this._transaction = mr.isTransaction(source);

        let expirationL = mr.getExpiration(source);
        if (expirationL == 0) {
            this._expiration = null;
        } else {
            this._expiration = new Date(expirationL);
        }
    }

    /**
     * 获取消息源
     */
    getSource() {
        return this._source;
    }

    getSender(): string {
        return this._sender;
    }

    /**
     * 跟踪ID
     */
    getTid(): string {
        return this._tid;
    }

    /**
     * 标签
     */
    getTag(): string {
        return this._tag;
    }

    /**
     * 主题
     */
    getTopic(): string {
        return this._topic;
    }

    /**
     * 消费者组
     * */
    getConsumerGroup(): string {
        return this._consumerGroup;
    }

    getContent(): string {
        return this.getBodyAsString();
    }

    getBody(): ArrayBuffer {
        return this._source.data().getArray()!;
    }

    /**
     * 内容
     */
    getBodyAsString(): string {
        return this._source.dataAsString();
    }

    /**
     * 质量等级（0 或 1）
     */
    getQos(): number {
        return this._qos;
    }

    getAttr( name:string):string|null {
        return this._source.meta(MqConstants.MQ_ATTR_PREFIX + name);
    }

    /**
     * 过期时间
     * */
    getExpiration(): Date | null {
        return this._expiration;
    }

    /**
     * 是否事务
     */
    isTransaction(): boolean {
        return this._transaction;
    }

    /**
     * 是否有序
     */
    isSequence(): boolean {
        return this._sequence;
    }

    /**
     * 已派发次数
     */
    getTimes(): number {
        return this._times;
    }

    /**
     * 回执
     */
    acknowledge(isOk: boolean) {
        this._clientInternal.reply(this._session, this._source, this, isOk, null);
    }

    response(entity: Entity | null) {
        this._clientInternal.reply(this._session, this._source, this, true, entity);
    }

    toString(): string {
        return "MqMessageReceived{" +
            "tid='" + this._tid + '\'' +
            ", tag='" + this._tag + '\'' +
            ", topic='" + this._topic + '\'' +
            ", body='" + this.getBodyAsString() + '\'' +
            '}';
    }
}

File: folkmq-lang-python/folkmq/folkmq/client/MqMessageReceived.py
from abc import abstractmethod
from datetime import datetime

from socketd.transport.core import Entity
from socketd.transport.core.Message import Message
from socketd.transport.core.Session import Session

from folkmq.client.MqMessage import MqMessageBase
from folkmq.common.MqConstants import MqConstants
from folkmq.common.MqTopicHelper import MqTopicHelper
from folkmq.common.MqUtils import MqUtils


class MqMessageReceived(MqMessageBase):
    # 主题
    @abstractmethod
    def get_topic(self) -> str:
        ...

    # 内容
    @abstractmethod
    def get_body_as_string(self) -> str:
        ...

    # 消费者组
    @abstractmethod
    def get_consumer_group(self) -> str:
        ...

    # 已派发次数
    @abstractmethod
    def get_times(self) -> int:
        ...

    # 回执
    @abstractmethod
    def acknowledge(self, isOk: bool):
        ...

    # 响应
    @abstractmethod
    def response(self, entity:Entity):
        ...

class MqMessageReceivedImpl(MqMessageReceived):
    def __init__(self, clientInternal:'MqClientInternal', session:Session, source:Message):
        self.__clientInternal = clientInternal
        self.__session = session
        self.__source = source

        mr  = MqUtils.get_of(source)

        self.__sender = mr.get_sender(source)

        self.__key = mr.get_key(source)
        self.__tag = mr.get_tag(source)
        self.__fullTopic = mr.get_topic(source)
        self.__topic = MqTopicHelper.get_topic(self.__fullTopic)
        self.__consumerGroup = mr.get_consumer_group(source)

        self.__qos = mr.get_qos(source)
        self.__times = mr.get_times(source)
        self.__sequence = mr.is_sequence(source)
        self.__broadcast = mr.is_broadcast(source)
        self.__transaction = mr.is_transaction(source)

        self.__expirationL = mr.get_expiration(source)
        self.__expiration = self.__expirationL

        self.__replied = False

    def is_replied(self) -> bool:
        return self.__replied

    def set_replied(self, replied: bool):
        self.__replied = replied

    def get_source(self) -> Message:
        return self.__source

    def get_sender(self) -> str | None:
        return self.__sender

    def get_key(self) -> str:
        return self.__key

    def get_tag(self) -> str:
        return self.__tag

    def get_topic(self) -> str:
        return self.__topic

    def get_full_topic(self)->str:
        return self.__fullTopic

    def get_consumer_group(self) -> str:
        return self.__consumerGroup

    def get_body(self) -> bytes:
        return self.__source.data_as_bytes()

    def get_body_as_string(self) -> str:
        return self.__source.data_as_string()

    def get_qos(self) -> int:
        return self.__qos

    def get_attr(self, name: str) -> str | None:
        return self.__source.meta(MqConstants.MQ_ATTR_PREFIX + name)

    def get_expiration(self) -> datetime | None:
        return self.__expiration

    def is_transaction(self) -> bool:
        return self.__transaction

    def is_sequence(self) -> bool:
        return self.__sequence

    def is_broadcast(self) -> bool:
        return self.__broadcast

    def get_times(self) -> int:
        return self.__times

    def acknowledge(self, isOk: bool):
        self.__clientInternal.reply(self.__session, self, isOk, None)

    def response(self, entity:Entity):
        self.__clientInternal.reply(self.__session, self, True, entity)

    def __str__(self) ->str:
        return "MqMessageReceived{" + \
            "key='" + self.__key + '\'' + \
            ", tag='" + self.__tag + '\'' + \
            ", topic='" + self.__topic + '\'' + \
            ", body='" + self.get_body_as_string() + '\'' + \
            '}'


